Dockerfile

```diff
@@ -25,27 +43,51 @@ ARG PKG_SRC
 ARG PKG_NAME
 
 WORKDIR /app
+
+ENV PIP_NO_COMPILE=1 \
+    PIP_DISABLE_PIP_VERSION_CHECK=1
+
+# Install build tools to compile dependencies that don't have prebuilt wheels
+RUN apk add --update --no-cache build-base git poetry libffi-dev
+RUN pip install --no-cache-dir --upgrade pip setuptools wheel
+
+# Copy the bare minimum needed for specifying dependencies.
+# This will enable better layer caching and faster builds when iterating locally.
+# `--without-hashes` is used to ensure the `pip` cache mount is used for packages that
+# would otherwise only match the sdist and therefore have to be built for every run.
+# References:
+#   * https://pythonspeed.com/articles/pipenv-docker/
+#   * https://hub.docker.com/r/docker/dockerfile
+COPY pyproject.toml poetry.lock ./
+RUN poetry export --without-hashes --format requirements.txt --output requirements.txt
+
+# Cache the pip installed dependencies
+# NOTE: This `--mount` feature requires BUILDKIT to be used
+RUN --mount=type=cache,id=pip,target=/root/.cache/pip \
+    set -eux; \
+    pip cache info; \
+    pip cache list; \
+    pip install --user -r requirements.txt
 COPY "${PKG_SRC:-.}" .
-RUN apt update \
-    && apt upgrade -y \
-    # Install build tools to compile dependencies that don't have prebuilt wheels
-    && apt install -y git build-essential \
-    && pip install --no-cache-dir --upgrade pip setuptools wheel \
-    && pip install --user --no-cache-dir ${PKG_NAME:-.}
+RUN pip install --user --no-cache-dir ${PKG_NAME:-.}
+RUN find /root/.local -type f -name '*.pyc' -delete
 
 # Explicitly specify a platform that is supported by `phylum-init`
-FROM --platform=linux/amd64 python:3.10-slim
+FROM --platform=linux/amd64 python:3.10-alpine
 
 LABEL maintainer="Phylum, Inc. <engineering@phylum.io>"
-# copy only Python packages to limit the image size
+
+# Copy only Python packages to limit the image size
 COPY --from=builder /root/.local /root/.local
-ENV PATH=/root/.local/bin:$PATH
-RUN apt update \
-    && apt upgrade -y \
-    && apt install -y git \
-    && apt clean \
-    && apt purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false \
-    && rm -rf /var/lib/apt/lists /var/cache/apt/archives \
-    && phylum-init
+
+ENV PATH=/root/.local/bin:$PATH \
+    PYTHONDONTWRITEBYTECODE=1
+
+RUN set -eux; \
+    apk add --update --no-cache git; \
+    phylum-init; \
+    find / -type f -name '*.pyc' -delete; \
+    # Ensure no keys used during image creation are leaked through
+    unset PHYLUM_API_KEY
```

Comment:
`unset` is a shell builtin that will remove the environment variable for the current process (i.e., this particular `RUN` command). I don't see how this could possibly have any impact on future layers.

Also, if the API key was set in a way that impacted future layers (i.e., the `ENV` directive), unsetting it wouldn't stop the leak (because all intermediate layers still exist in the final image).

An alternative approach would be to check for the environment variable and fail the build if it's set. I think that would look something like this:

```
RUN test -z "${PHYLUM_API_KEY:-}" || exit 1
```

But I think that is probably more paranoid than we need to be... The only thing it would catch is if someone adds `ENV PHYLUM_API_KEY=...` to this file above that line. And anything like that should immediately fail code review.

Reply:
Ah...good catch...it has been removed.
This was a holdover from a time when, after creating the image, I ran it as a container and viewed the environment variables with the `env` command. `PHYLUM_API_KEY` was there...but I didn't realize at the time that:
* it was only there b/c I used the `docker run -it --rm -e PHYLUM_API_KEY --mount type=bind,src=$(pwd),dst=/phylum -w /phylum phylumio/phylum-ci` form to run the container and `PHYLUM_API_KEY` was set in my local/host environment
* `ENV` directives in the Dockerfile for the *last* build stage are the only ones that carry forward to the resulting image